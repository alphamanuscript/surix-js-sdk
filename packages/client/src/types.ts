import { WrappedEntity, WrappedFile } from '@surix/data-helpers';

export interface Project {
  entities: ProjectEntities;
  files: ProjectFiles;
  tags: ProjectTags;
}

export interface ProjectEntities {
  get(id: string): Promise<WrappedEntity>;
  query(query?: Query): Promise<WrappedEntity[]>;
  create(entity: Entity): Promise<WrappedEntity>;
  put(entity: Entity): Promise<WrappedEntity>;
}

export interface ProjectFiles {
  get(id: string): Promise<WrappedFile>;
  list(): Promise<WrappedFile[]>;
}

export interface ProjectTags {
  list(): Promise<TagList>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];

export enum FieldType {
  TEXT = 'text',
  BOOLEAN = 'boolean',
  NUMBER = 'number',
  DATETIME = 'datetime',
  PHONE = 'phone',
  EMAIL = 'email',
  FILE = 'file',
  OBJECT = 'object',
  LIST = 'list'
}

export interface FieldPair {
  type: FieldType;
  value: RawEntityData | string | RawEntityData[] | number;
}

export interface RawEntityField extends FieldPair {
}

export interface RawEntityData {
  [fieldName: string]: RawEntityField;
}

/**
 * this interface describe the shape of the entity
 * as used within the system and the API
 */
export interface RawEntity {
  data: RawEntityData;
  tags?: string[];
}

export interface DataField {
  [fieldName: string]: string | number | object | boolean;
}
export interface Entity {
  data: DataField;
  tags?: string[];
  _id?: string;
}