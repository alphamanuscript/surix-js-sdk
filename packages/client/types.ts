import { EntityWrapper } from '@surix/data-helpers';

export interface Project {
  db: ProjectDatabase;
  files: ProjectFiles;
}

export interface ProjectDatabase {
  get(id: string): Promise<EntityWrapper>;
  query(query: Query): Promise<EntityWrapper[]>;
  getTags(): Promise<TagList[]>;
}

export interface ProjectFiles {
  get(id: string): Promise<ApiFile>;
  list(): Promise<ApiFile[]>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface ApiFile {
  name: string;
  status: 'failed' | 'ready' | 'pending';
  public: boolean;
  downloadUrl: string;
  tags: TagList[];
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];