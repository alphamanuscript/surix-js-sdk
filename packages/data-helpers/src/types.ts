export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldPathArray = Array<string|number>;

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | FileValue | ListValue | ObjectValue;

export interface FileValue {
    ref: string;
    mimeType: string;
    downloadUrl: string;
}

export type ListValue = ListItemValue[];

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: Field;
}

export interface Field extends TypeValuePair {
    name: string;
    label: string;
}

export interface TypeValuePair {
    type: FieldType;
    value: FieldValue;
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}

export type FieldOrValuePair = Field | TypeValuePair;
export type NullableFieldOrValuePair = FieldOrValuePair | undefined;

export interface WrappedEntity {
    _id: string;
    id: string;
    createdAt: Date;
    updatedAt: Date;
    createdBy: AuthContext;
    tags: string[];
    rawEntity: ApiEntity;
    /**
     * returns the plain value of the field at the specified path
     * @param fieldPath the key of the field, this can be a
     * dotted string "field.nestedField" or an array ["field", "nestedField"]
     * the path can include field names as well as array indices
     * @param defaultValue 
     */
    get (fieldPath: string | FieldPathArray, defaultValue?: any): any;
    /**
     * returns the type of the field with the specified key
     * @param fieldPath the key of the field, this can be a
     * dotted string "field.nestedField" or an array ["field", "nestedField"]
     * the path can include field names as well as array indices
     * @param defaultType the type to return if the entity does not have the field
     */
    type (fieldPath: string | FieldPathArray, defaultType?: FieldType): FieldType|undefined;
    /**
     * returns the label of the field with the specified key
     * @param fieldPath the key of the field, this can be a
     * dotted string "field.nestedField" or an array ["field", "nestedField"]
     * the path can include field names as well as array indices
     * @param defaultLabel the label to return if the field does not exist
     */
    label (fieldPath: string | FieldPathArray, defaultLabel?: string): string|undefined;
    /**
     * returns the raw field at the specified key
     * @param fieldPath the key of the field, this can be a
     * dotted string "field.nestedField" or an array ["field", "nestedField"]
     * the path can include field names as well as array indices
     * @param defaultField the type to return if the entity does not have the field
     */
    field (fieldPath: string | FieldPathArray, defaultField?: FieldOrValuePair):
        NullableFieldOrValuePair;
    /**
     * returns the entity data
     * as plain (deflated) key-value pair
     */
    data (): object;
}

export interface ApiFile {
    _id: string;
    name: string;
    filename: string;
    status: 'pending' | 'failed' | 'ready';
    size: number;
    mimeType: string;
    public: boolean;
    tags: string[];
    uploadUrl: string;
    downloadUrl: string;
    createdBy: AuthContext;
    createdAt: string;
    updatedAt: string;
}

export interface WrappedFile {
    _id: string;
    id: string;
    name: string;
    filename: string;
    status: 'pending' | 'failed' | 'ready';
    size: number;
    mimeType: string;
    public: boolean;
    tags: string[];
    uploadUrl: string;
    downloadUrl: string;
    createdBy: AuthContext;
    createdAt: Date;
    updatedAt: Date;
    rawFile: ApiFile;
}