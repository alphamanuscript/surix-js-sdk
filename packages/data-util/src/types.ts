export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | ListItemValue[] | ObjectValue;

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: {
        name: string;
        type: FieldType;
        label: string;
        value: FieldValue;
    };
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}