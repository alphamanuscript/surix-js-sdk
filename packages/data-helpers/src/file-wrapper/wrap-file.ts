import { ApiFile, WrappedFile } from '../types';

export function wrapFile (file: ApiFile): WrappedFile {
  return {
    ...file,
    id: file._id,
    createdAt: new Date(file.createdAt),
    updatedAt: new Date(file.updatedAt),
    get rawFile (): ApiFile {
      return file;
    }
  };
}

export function wrapFileArray (files: ApiFile[]): WrappedFile[] {
  return files.map(wrapFile);
}