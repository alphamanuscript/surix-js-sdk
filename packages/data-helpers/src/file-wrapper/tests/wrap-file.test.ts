import { ApiFile } from '../../types';
import { wrapFile, wrapFileArray} from '../wrap-file';

describe('wrapFile', () => {
  let file: ApiFile;
  beforeEach(() => {
    file = {
      _id: 'f1',
      name: 'My File',
      filename: 'f1.jpeg',
      mimeType: 'image/jpeg',
      size: 246,
      status: 'ready',
      downloadUrl: 'http://download/f1',
      uploadUrl: 'http://upload/f1',
      public: true,
      tags: ['t1', 't2'],
      createdBy: {
        _id: 'f1',
        type: 'user'
      },
      createdAt: '2017-02-22T09:19:33.885Z',
      updatedAt: '2018-10-22T09:19:33.885Z',
    };
  });
  const wrap = () => wrapFile(file);
  it('should transform api file into WrappedFile', () => {
    expect(wrap()).toMatchSnapshot();
  });
  it('should return timestamps as actual Dates', () => {
    const wrapped = wrap();
    expect(wrapped.createdAt).toBeInstanceOf(Date);
    expect(wrapped.updatedAt).toBeInstanceOf(Date);
  });
  it('should have rawFile property that returns original ApiFile', () => {
    expect(wrap().rawFile).toEqual(file);
  });
});

describe('wrapFileArray', () => {
  it('should transform array of ApiFile into array of WrappedFile', () => {
    const files: ApiFile[] = [
      {
        _id: 'f1',
        name: 'My File',
        filename: 'f1.jpeg',
        mimeType: 'image/jpeg',
        size: 246,
        status: 'ready',
        downloadUrl: 'http://download/f1',
        uploadUrl: 'http://upload/f1',
        public: true,
        tags: ['t1', 't2'],
        createdBy: {
          _id: 'u1',
          type: 'user'
        },
        createdAt: '2017-02-22T09:19:33.885Z',
        updatedAt: '2018-10-22T09:19:33.885Z',
      },
      {
        _id: 'f2',
        name: 'My Other File',
        filename: 'f2.pdf',
        mimeType: 'application/pdf',
        size: 400,
        status: 'pending',
        downloadUrl: '',
        uploadUrl: 'http://upload/f2',
        public: false,
        tags: ['t1', 't2'],
        createdBy: {
          _id: 'u2',
          type: 'user'
        },
        createdAt: '2017-02-22T09:19:33.885Z',
        updatedAt: '2018-10-22T09:19:33.885Z',
      }
    ];
    const wrapped = wrapFileArray(files);
    expect(wrapped).toMatchSnapshot();
  });
});
