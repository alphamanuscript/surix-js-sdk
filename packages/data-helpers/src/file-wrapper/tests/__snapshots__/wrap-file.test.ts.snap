// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapFile should transform api file into WrappedFile 1`] = `
Object {
  "_id": "f1",
  "createdAt": 2017-02-22T09:19:33.885Z,
  "createdBy": Object {
    "_id": "f1",
    "type": "user",
  },
  "downloadUrl": "http://download/f1",
  "filename": "f1.jpeg",
  "id": "f1",
  "mimeType": "image/jpeg",
  "name": "My File",
  "public": true,
  "rawFile": Object {
    "_id": "f1",
    "createdAt": "2017-02-22T09:19:33.885Z",
    "createdBy": Object {
      "_id": "f1",
      "type": "user",
    },
    "downloadUrl": "http://download/f1",
    "filename": "f1.jpeg",
    "mimeType": "image/jpeg",
    "name": "My File",
    "public": true,
    "size": 246,
    "status": "ready",
    "tags": Array [
      "t1",
      "t2",
    ],
    "updatedAt": "2018-10-22T09:19:33.885Z",
    "uploadUrl": "http://upload/f1",
  },
  "size": 246,
  "status": "ready",
  "tags": Array [
    "t1",
    "t2",
  ],
  "updatedAt": 2018-10-22T09:19:33.885Z,
  "uploadUrl": "http://upload/f1",
}
`;

exports[`wrapFileArray should transform array of ApiFile into array of WrappedFile 1`] = `
Array [
  Object {
    "_id": "f1",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "http://download/f1",
    "filename": "f1.jpeg",
    "id": "f1",
    "mimeType": "image/jpeg",
    "name": "My File",
    "public": true,
    "rawFile": Object {
      "_id": "f1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "downloadUrl": "http://download/f1",
      "filename": "f1.jpeg",
      "mimeType": "image/jpeg",
      "name": "My File",
      "public": true,
      "size": 246,
      "status": "ready",
      "tags": Array [
        "t1",
        "t2",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f1",
    },
    "size": 246,
    "status": "ready",
    "tags": Array [
      "t1",
      "t2",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f1",
  },
  Object {
    "_id": "f2",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u2",
      "type": "user",
    },
    "downloadUrl": "",
    "filename": "f2.pdf",
    "id": "f2",
    "mimeType": "application/pdf",
    "name": "My Other File",
    "public": false,
    "rawFile": Object {
      "_id": "f2",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u2",
        "type": "user",
      },
      "downloadUrl": "",
      "filename": "f2.pdf",
      "mimeType": "application/pdf",
      "name": "My Other File",
      "public": false,
      "size": 400,
      "status": "pending",
      "tags": Array [
        "t1",
        "t2",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f2",
    },
    "size": 400,
    "status": "pending",
    "tags": Array [
      "t1",
      "t2",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f2",
  },
]
`;
